### ISSUE TITLE
incremental kompilation

### ISSUE THREAD

**Issue by cos** [kframework/k-legacy#505]
I'm adding this as a feature request before I get used to the status quo. 

The current time to compile even the simplest definition is ridiculous. I understand that there are efforts to make this faster on multiple fronts. But as fast as it would get, it would still be far from instant.

The solution in my opinion is incremental kompilation. If kompile would only have to parse the definition which have changed (and their transitive closure), I think the kompilation would be must faster in most instances, even with the current implementations.

**Comment by chathhorn** [comment]
I'll second this. The C semantics takes ~20 minutes to kompile from clean for the whole thing on fslwork (and probably about a year on my laptop). Which wouldn't be quite so frustrating if kompile would report more than a single error before crapping out, but I suppose that's a separate issue.

**Comment by osa1** [comment]
Slow kompilation times are discussed several times at meetings in the past. There are lots of low-hanging fruits that will make kompilation a lot faster but last time I tried to implement `kompile` source code was a huge obstacle for me. If with recent refactorings it got better then with some simple changes it can be made a lot faster. (like serializing parsed sources with last modification date and only to reparse when source is changed etc.)

**Comment by grosu** [comment]
Guys, this has been discussed a LOT in the past.  Particularly, the incremental kompilation.  The real question is: who's volunteering to work on it?  Everybody seems to be overly busy with their stuff.  Those of you who have already proposed solutions to this, or even worked for a while on this problem, can you please list your thoughts/findings here, so whoever will work on this will have a head-start?  Thanks!

**Comment by chathhorn** [comment]
Sorry -- it's actually more satisfying and lot less work just to complain.

**Comment by andreistefanescu** [comment]
In my experience, one does not change the syntax or the configuration very often. So I propose to keep a kast version of every rule cached, and only re-parse the rules that change. This would speed up parsing a lot, at least for CIL and other languages which use concrete syntax.

**Comment by radumereuta** [comment]
Whoever finds the motivation to work on this, should come find me.
The most speed boost that I can think of, can be done during file/module gathering
(see org.kframework.parser.generator.BasicParser)
Saving and restoring parts of the file (modules, rules, syntax declarations) is actually really simple. The trickier part is invalidating the cached information. A graph of dependencies needs to be maintained and updated each time a modification is done.

**Comment by cos** [comment]
#559 solves this issues for many common cases.

**Comment by radumereuta** [comment]
Should be fixed by passing #559 
At the moment it's blocked by a weird bug in krun which I wasn't able to track down, so I asked @dwightguth to have a look.
